### ISSUE TITLE
False positive for switch expressions in MissingSwitchDefault

### ISSUE THREAD

**Issue by wlmim** [checkstyle/checkstyle#16599]
I have read check documentation: https://checkstyle.org/checks/xxxxxx/nameofaffectedcheck.html
I have downloaded the latest checkstyle from: https://checkstyle.org/cmdline.html#Download_and_Run
I have executed the cli and showed it below, as cli describes the problem better than 1,000 words

```bash
/var/tmp $ javac mytest\SwitchExpressionCheckstyle.java

/var/tmp $ cat checkstyle_config.xml
<?xml version="1.0" encoding="UTF-8"?>
<!DOCTYPE module PUBLIC "-//Puppy Crawl//DTD Check Configuration 1.3//EN" "http://www.puppycrawl.com/dtds/configuration_1_3.dtd">

<module name="Checker">
  <module name="TreeWalker">
    <module name="MissingSwitchDefault"/>
  </module>
</module>



/var/tmp $ cat mytest\SwitchExpressionCheckstyle.java
package mytest;

/**
 * My Class.
 */
public class SwitchExpressionCheckstyle {

    private int something = 0;

    /**
     * Do a switch.
     * 
     * @param s
     *            The enum value to switch on.
     */
    public void doSwitch(SomeEnum s) {
        switch (s) {
            // Do nothing
            case D -> {}
            case R -> this.something += 5;
            case T -> this.something += 7;
        }
        System.out.println(something);
    }

    /**
     * Some enum to switch on.
     */
    public static enum SomeEnum {
        /** T. */
        T,
        /** R. */
        R,
        /** D. */
        D;
    }
}


/var/tmp $ RUN_LOCALE="-Duser.language=en -Duser.country=US"
/var/tmp $ java $RUN_LOCALE -jar checkstyle-checkstyle-10.21.4-all.jar -c checkstyle_config.xml mytest\SwitchExpressionCheckstyle.java
Starting audit...
[WARN] C:\...\mytest\SwitchExpressionCheckstyle.java:17:9: switch without "default" clause. [MissingSwitchDefault]
Audit done.
```

---

I expected the switch expression not to trigger the `MissingSwitchDefault` check.
This seems like a regression of [#10866](https://github.com/checkstyle/checkstyle/issues/10866)

---

**Comment by mahfouz72** [comment]
This is not a switch expression. This is a switch statement using the arrow (`->`) syntax. The compiler does not enforce exhaustiveness in this case, so a violation is expected.

**Comment by wlmim** [comment]
Oh, my bad, you're right of course. The switch statement merely gives a compiler warning, not an error.

So my next thought would be to disable this check as in our case, a compiler warning (when there's really an enum value missing) is as good as a checkstyle warning (and the checkstyle warning is a false positive if all enum values are covered). However, removing the MissingSwitchDefault check would open the door to ppl forgetting the default case for switching over anything but enums.

Would it be possible to add a configuration option in MissingSwitchDefault to allow omitting the default case when switching over enums? The rationale being that the compiler would emit a warning. As an option, it could be backwards-compatible by default.

**Comment by mahfouz72** [comment]
See the Rationale of the check:

> Rationale: It's usually a good idea to introduce a default case in every switch statement. Even if the developer is sure that all currently possible cases are covered, this should be expressed in the default branch, e.g. by using an assertion. This way the code is protected against later changes

Suppose this enum is extended with new cases in the future, the default branch will act as a safeguard. Therefore, even if all enum values are currently covered, it is still recommended to include a default case. You may agree or disagree with this approach, but this check enforces it by design.


> Would it be possible to add a configuration option in MissingSwitchDefault to allow omitting the default case when switching over enums?

Also FYI, this may not be even possible because Checkstyle does not have access to type information. Checkstyle operates on the Abstract Syntax Tree without semantic analysis.  See [limitations](https://checkstyle.sourceforge.io/writingchecks.html#Limitations). So, it can not reliably determine whether a switch is based on an enum

**Comment by wlmim** [comment]
Okay, thanks for taking the time to clarify.

I ended up implementing a default branch and throwing an exception. I'll close this issue now.
